x/reports/keeper/msg_server.go

```diff
@@ -0,0 +1,284 @@
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// Check if the subspace exists
+	if !k.HasSubspace(ctx, msg.SubspaceID) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "subspace with id %d not found", msg.SubspaceID)
+	}
+
+	// Check if the reason exists
+	if !k.HasReason(ctx, msg.SubspaceID, msg.ReasonID) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "reason with id %d not found inside subspace %d", msg.ReasonID, msg.SubspaceID)
+	}
+
+	// Check the permission to report
+	if !k.HasPermission(ctx, msg.SubspaceID, msg.Reporter, subspacestypes.PermissionReportContent) {
+		return nil, sdkerrors.Wrap(subspacestypes.ErrPermissionDenied, "you cannot report content inside this subspace")
+	}
+
+	// Get the next report id
+	reportID, err := k.GetNextReportID(ctx, msg.SubspaceID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create and validate the report
+	report := types.NewReport(
+		msg.SubspaceID,
+		reportID,
+		msg.ReasonID,
+		msg.Message,
+		msg.Target.GetCachedValue().(types.ReportTarget),
+		msg.Reporter,
+		ctx.BlockTime(),
+	)
+	err = k.ValidateReport(ctx, report)
+	if err != nil {
+		return nil, err
+	}
+
+	// Store the report
+	k.SaveReport(ctx, report)
+
+	// Update the id for the next report
+	k.SetNextReportID(ctx, msg.SubspaceID, report.ID+1)
+
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			sdk.EventTypeMessage,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
+			sdk.NewAttribute(sdk.AttributeKeyAction, sdk.MsgTypeURL(msg)),
+			sdk.NewAttribute(sdk.AttributeKeySender, msg.Reporter),
+		),
+		sdk.NewEvent(
+			types.EventTypeCreateReport,
+			sdk.NewAttribute(types.AttributeKeySubspaceID, fmt.Sprintf("%d", msg.SubspaceID)),
+			sdk.NewAttribute(types.AttributeKeyReportID, fmt.Sprintf("%d", report.ID)),
+			sdk.NewAttribute(types.AttributeKeyReporter, msg.Reporter),
+			sdk.NewAttribute(types.AttributeKeyCreationTime, report.CreationDate.Format(time.RFC3339)),
+		),
+	})
+
+	return &types.MsgCreateReportResponse{
+		ReportID:     report.ID,
+		CreationDate: report.CreationDate,
+	}, nil
+}
+
+// DeleteReport defines the rpc method for Msg/DeleteReport
+func (k msgServer) DeleteReport(goCtx context.Context, msg *types.MsgDeleteReport) (*types.MsgDeleteReportResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// Check if the subspace exists
+	if !k.HasSubspace(ctx, msg.SubspaceID) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "subspace with id %d not found", msg.SubspaceID)
+	}
+
+	// Check if the report exists
+	report, found := k.GetReport(ctx, msg.SubspaceID, msg.ReportID)
+	if !found {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "report with id %d not found inside subspace %d", msg.ReportID, msg.SubspaceID)
+	}
+
+	// Check the permission to delete reports
+	isModerator := k.HasPermission(ctx, msg.SubspaceID, msg.Signer, subspacestypes.PermissionManageReports)
```

Comment:
It is weird to me that moderator can delete the report. Can you elaborate the usecase?